File: java/FirstMissingPositive.java
public class FirstMissingPositive {
    public int firstMissingPositive(int[] num) {
        if (num == null) {
            return 1;
        }
        int n = num.length;
        if (n == 0) {
            return 1;
        }
        for (int i = 0; i < n; ++i) {
            // make sure number in range [1, n] in right position
            while (1 <= num[i] && num[i] <= n 
                    && i + 1 != num[i] 
                    // avoid such case [1, 1]
                    && num[i] != num[num[i] - 1]) {
                int temp = num[num[i] - 1];
                num[num[i] - 1] = num[i];
                num[i] = temp;
            }
        }
        for (int i = 0; i < n; ++i) {
            if (num[i] != i + 1) {
                return i + 1;
            }
        }
        return n + 1;
    }

    public static void main(String[] args) {
        FirstMissingPositive solver = new FirstMissingPositive();
        int[] num = null;

        num = new int[] {3, 4, -1, 1};
        System.out.println(solver.firstMissingPositive(num));

        num = new int[] {1, 1};
        System.out.println(solver.firstMissingPositive(num));
    }
}

File: python/first_missing_positive.py
#!/usr/bin/env python3

class Solution:
    def firstMissingPositive(self, A):
        n = len(A)
        for i in range(n):
            while 1 <= A[i] and A[i] <= n and A[i] != i + 1 and A[i] != A[A[i] - 1]:
                [A[A[i] - 1], A[i]] = [A[i], A[A[i] - 1]]
        res = 1
        for i in range(n):
            if i != A[i] - 1:
                break;
            res = res + 1
        return res

def main():
    solver = Solution()
    A = [1, 2, 0]
    print(solver.firstMissingPositive(A))
    A = [3, 4, 1, -1]
    print(solver.firstMissingPositive(A))

if __name__ == '__main__':
    main()
